### PR TITLE
Recorders

### PR DESCRIPTION
**Pull request by Robert-Kolmos**
Now a fast forward from #45
- Finish implementing RecorderManager and having it start and stop recorders.
- Setup DeviceMotionJsonRecorder to write json data to files and supply FileResults

### CODE REVIEW

File: presentation/src/main/java/org/sagebionetworks/research/presentation/inject/RecorderConfigPresentationModule.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import dagger.MapKey;
+import dagger.Module;
+import dagger.Provides;
+import dagger.multibindings.IntoMap;
+import dagger.multibindings.StringKey;
+
+@Module
+public abstract class RecorderConfigPresentationModule {
+    public interface RecorderConfigPresentationFactory {
+        RecorderConfigPresentation create(RecorderConfiguration configuration);
```
Comment (liujoshua) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/inject/RecorderConfigPresentationModule.java]: This interface is used outside of Dagger, so it is important enough to live outside of a Dagger module

File: presentation/src/main/java/org/sagebionetworks/research/presentation/perform_task/PerformTaskViewModel.java [hunk 2]
```diff
@@ -318,6 +340,10 @@ protected void updateCurrentStep(@Nullable Step nextStep, @NonNull TaskResult ta
     void handleTaskLoad(Task task) {
         LOGGER.debug("Loaded task: {}", task);
         this.task = task;
+        this.recorderManager = new RecorderManager(this.task, this.taskRunUuid, this.getApplication(), this.recorderFactory,
+                this.recorderConfigPresentationFactory);
+        // Subscribe to the recorder results and put them in the async results.
+        this.compositeDisposable.add(Flowable.create(this.recorderManager, BackpressureStrategy.BUFFER).subscribe(this::addAsyncResult));
```
Comment (liujoshua) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/perform_task/PerformTaskViewModel.java]: I think if the UI goes away before the io is done, we lose the data. Add a TODO

File: presentation/src/main/java/org/sagebionetworks/research/presentation/recorder/data/DataLogger.java [hunk 3]
```diff
@@ -75,40 +79,45 @@
...
+        // append either the start or deliminator for separation
+        String outputString = (isFirstJsonObject ? this.start : this.deliminator) + next;
+        this.outputStream.print(outputString);
         this.isFirstJsonObject = false;
     }
 
     @Override
     public void onError(final Throwable t) {
         this.outputStream.close();
+        t.printStackTrace();
```
Comment (liujoshua) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/recorder/data/DataLogger.java]: remove

File: presentation/src/main/java/org/sagebionetworks/research/presentation/recorder/device_motion/DeviceMotionJsonAdapter.java [hunk 4]
```diff
@@ -118,32 +118,35 @@ public static JsonObject createJsonObject(ReactiveSensorEvent event) {
...
-            case Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR:
-                recordRotationVectorEvent(sensorEvent, jsonObject);
-                break;
-            case Sensor.TYPE_ACCELEROMETER_UNCALIBRATED:
-            case Sensor.TYPE_GYROSCOPE_UNCALIBRATED:
-            case Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED:
-                recordUncalibratedEvent(sensorEvent, jsonObject);
-                break;
+        Sensor sensor = sensorEvent.sensor;
+        if (sensor != null) {
```
Comment (liujoshua) [inline presentation/src/main/java/org/sagebionetworks/research/presentation/recorder/device_motion/DeviceMotionJsonAdapter.java]: This means we record a blob with just the common elements?

File: presentation/src/main/java/org/sagebionetworks/research/presentation/perform_task/PerformTaskViewModel.java (hunk 2 above)
Comment (Robert-Kolmos) [reply presentation/src/main/java/org/sagebionetworks/research/presentation/perform_task/PerformTaskViewModel.java]: Is this true, I thought the purpose of a view model was to outlive lifecycle states.

File: domain/src/main/java/org/sagebionetworks/research/domain/step/interfaces/Step.java [hunk 5]
```diff
@@ -53,6 +59,13 @@
     @NonNull
     Step copyWithIdentifier(@NonNull String identifier);
 
+    /**
+     * Returns the AsyncActionConfigurations for this step.
+     * @return the AsyncActionConfigurations for this step.
+     */
+    @NonNull
```
[the remark this answers is not in the document]
Comment (Robert-Kolmos) [reply domain/src/main/java/org/sagebionetworks/research/domain/step/interfaces/Step.java]: Hmm I thought it did, perhaps I am mistaken, I can refactor that to only support async actions on task and section steps in a later pull request if this is the case.

File: data/src/main/java/org/sagebionetworks/research/data/ResourceTaskRepository.java [hunk 6]
```diff
@@ -189,6 +195,63 @@ private InputStreamReader getAsset(String assetPath) throws IOException {
         return new InputStreamReader(assetManager.open(assetPath), UTF_8);
     }
 
+    private static Set<AsyncActionConfiguration> getAsyncActions(Task task) {
+        return ResourceTaskRepository.getAsyncActionsHelper(task.getSteps(), new HashSet<>(task.getAsyncActions()));
+    }
+
+    private static Set<AsyncActionConfiguration> getAsyncActionsHelper(List<Step> steps, Set<AsyncActionConfiguration> accumlator) {
```
[the remark this answers is not in the document]
Comment (Robert-Kolmos) [reply data/src/main/java/org/sagebionetworks/research/data/ResourceTaskRepository.java]: What about the traversal of through the section steps to accumulate all of the async actions. If this should also be done in presentation I would suggest making a Presentation layer Task object that can do this processing when it is initialized.

**Pull request merged**